Add Offline notification.

zutigrm opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #3820 

## Relevant technical choices

Deviations from IB:
* The provided url for connectivity check, can't be used, due to the CORS issue, so for best accuracy and avoiding problems that can occur in the future with the 3rd party services that are outside of our control, I used the existing healcheck endpoint, for an additional connection confirmation.
* The proposed usage of `SettingsNotification` component can't be used, as dismissal is making a request to the server, which cannot happen in offline state. So I used only needed parts of that component to assemble the notification and handled dismissal with component level state.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on assets/js/hooks/useMonitorInternetConnection.js:
```diff
@@ -0,0 +1,108 @@
+			: intervalValues.isOffline;
+	}
+
+	useInterval( async () => {
+		await checkInternetConnection();
+	}, intervalTime );
```
Similarly to my comment above, I think we could probably avoid making this async, and just pass `checkInternetConnection()` in directly here:

```suggestion
	useInterval( checkInternetConnection, intervalTime );
```

techanvil commented on assets/js/hooks/useMonitorInternetConnection.js:
```diff
@@ -0,0 +1,108 @@
+	}, [ setValue ] );
+
+	useLifecycles(
+		async () => {
+			global.addEventListener( 'online', checkInternetConnection );
+			global.addEventListener( 'offline', checkInternetConnection );
+
+			await checkInternetConnection();
```
Hmm, unless I am missing something, it doesn't look like there is any need to make this function async and await for `checkInternetConnection()`, as it doesn't need to do anything after `checkInternetConnection()` has run. I don't think that making it async will have any effect on how it's treated by `useLifecycles()` (and thus React)...

```suggestion
		() => {
			global.addEventListener( 'online', checkInternetConnection );
			global.addEventListener( 'offline', checkInternetConnection );

			checkInternetConnection();
```

techanvil commented on assets/js/hooks/useMonitorInternetConnection.js:
```diff
@@ -0,0 +1,108 @@
+
+			// We are only intersted if the request was successfull, to
+			// confirm online status.
+			const canReachInternetURL = onlineResponse.checks;
+
+			if ( canReachInternetURL ) {
+				setValue( 'isOnline', true );
+			}
+
+			if ( ! canReachInternetURL ) {
+				setValue( 'isOnline', false );
+			}
```
We could simplify this a bit (or further, combining to a single line):

```suggestion
			const canReachInternetURL = !!onlineResponse.checks;

			setValue( 'isOnline', canReachInternetURL );
```

techanvil commented on assets/js/hooks/useMonitorInternetConnection.test.js:
```diff
@@ -0,0 +1,169 @@
+	let store;
+
+	const mockOnlineStatus = ( status = true ) => {
+		Object.defineProperty( navigator, 'onLine', {
```
It would be good to clean this up, either in an `afterEach()` or an `afterAll()`, to avoid potentially impacting other tests. For example, here's how the `mockLocation()` utility handles cleanup:

https://github.com/google/site-kit-wp/blob/7ca132846658f1fc581a292429271711948ff822/tests/js/mock-browser-utils.js#L28-L44

techanvil commented on assets/js/hooks/useMonitorInternetConnection.js:
```diff
@@ -0,0 +1,108 @@
+ * @param {number}     intervalValues.isOnline  Interval when user is online.
+ * @param {number}     intervalValues.isOffline Interval when user is offline.
+ */
+export function useMonitorInternetConnection( intervalValues = null ) {
```
I'd suggest that we _don't_ pass in `intervalValues`, as it's not used in the production code. It doesn't really add much value testing the code using `intervalValues` because it's exercising an execution path that simply won't run in the real world.

Alternatively, make the argument required and do use it in production, then the tests as they stand will make more sense.

techanvil commented on assets/sass/components/settings/_googlesitekit-settings-notice.scss:
```diff
@@ -209,6 +209,30 @@
+.googlesitekit-settings-notice-offline-notice {
+	align-items: center;
+	bottom: 0;
+	gap: 15px;
```
I think this should be `$grid-gap-phone` for consistency with the notice padding:

```suggestion
	gap: $grid-gap-phone;
```

techanvil reviewed (changes requested):
Hi @zutigrm, this is looking good. I have left a few comments, please take a look.

zutigrm commented on assets/js/components/notifications/OfflineNotification.test.js:
```diff
@@ -0,0 +1,105 @@
+	it( 'should display the notification when offline', async () => {
+		registry.dispatch( CORE_UI ).setValue( 'isOnline', false );
+
+		const { container, waitForRegistry } = render(
```
I used `queryByText` to be able to use for checking if it is not in the document as well, to avoid error thrown by `getByText` when text is not present

zutigrm commented on assets/js/components/notifications/OfflineNotification.test.js:
```diff
@@ -0,0 +1,105 @@
+			)
+		).not.toBeNull();
+
+		// eslint-disable-next-line require-await
+		await act( async () => {
```
Indeed, originally I needed it for side-effects, but seems fine without it now

zutigrm commented on assets/js/googlesitekit/datastore/ui/ui.js:
```diff
@@ -33,6 +33,7 @@ const SET_VALUE = 'SET_VALUE';
 
 export const initialState = {
 	useInViewResetCount: 0,
+	isOnline: true,
```
Thanks for pointing that out. I agree, we should be consistent, I will include it

zutigrm commented:
@techanvil Thank you for the feedback. PR updated

techanvil commented on assets/js/hooks/useMonitorInternetConnection.test.js:
```diff
@@ -127,43 +139,4 @@ describe( 'useMonitorInternetConnection', () => {
 		expect( fetchMock ).toHaveFetched( healthCheckEndpoint );
 	} );
-
-	it( 'should check online status in correct intervals', async () => {
```
Hmm, I would have thought you could keep this test and rewrite it to use Jest [timer mocks](https://jestjs.io/docs/timer-mocks)...

zutigrm replied:
Since I removed the optional parameter for `intervalTime`, default timeout values are too hight to be able to test them, as we have 5000ms timeout limit, so `advanceTimersByTime` would timeout due to this.

techanvil replied:
Using the timer mocks decouples the tests from the real system time, allowing us to test arbitrarily high delays - you should be OK here. For example:

https://github.com/google/site-kit-wp/blob/274839bd27eebaad2d9c4a689beb4043449e80b1/assets/js/googlesitekit/datastore/user/surveys.test.js#L238-L244

techanvil commented on assets/js/components/notifications/OfflineNotification.test.js:
```diff
@@ -84,22 +79,15 @@ describe( 'InternalServerError', () => {
-		await act( async () => {
-			registry.dispatch( CORE_UI ).setValue( 'isOnline', true );
-		} );
+			queryByText( /you are currently offline/i )
```
This is arguably splitting hairs a bit, but we do prefer to use `getByText()` when we do expect the text to exist. Having it throw an error is fine as it will fail the test, which is what we want, and using `getByText()` is a little more specific, we can see from its usage that we definitely expect the text to exist and it will retain its intent if the code is modified.

techanvil reviewed (changes requested):
Thanks @zutigrm, just a few last comments to address and this will be good to go.

zutigrm commented on assets/js/googlesitekit/datastore/ui/ui.js:
```diff
@@ -58,6 +58,21 @@ export const actions = {
+	 * @param {boolean} value `isOnline` status.
+	 * @return {Object} Redux-style action.
+	 */
+	*setIsOnline( value ) {
+		invariant( isBoolean( value ), 'value must be boolean.' );
+
+		return yield actions.setValue( 'isOnline', value );
+	},
```
Thanks, makes sense

zutigrm commented on assets/js/googlesitekit/datastore/ui/ui.js:
```diff
@@ -58,6 +58,21 @@ export const actions = {
+	 * @param {boolean} value `isOnline` status.
+	 * @return {Object} Redux-style action.
+	 */
+	*setIsOnline( value ) {
```
Thank you, it slipped from my mind 🤦

techanvil reviewed (changes requested):
Thanks for the updates, @zutigrm - please see this [comment](https://github.com/google/site-kit-wp/pull/8033#discussion_r1444521968) I left while it was in execution regarding the use of timer mocks.

zutigrm commented:
@techanvil Thank you. I updated the tests

techanvil commented on assets/js/hooks/useMonitorInternetConnection.test.js:
```diff
@@ -0,0 +1,177 @@
+
+	it( 'should check online status in correct interval when online', () => {
+		jest.useFakeTimers();
+		jest.spyOn( global, 'setInterval' );
```
Hi @zutigrm, checking for calls to `setInterval()` is a bit too coupled to the implementation. I would suggest taking the approach of testing the observable side effects instead. For example we could do something like this:

```js
	it( 'should check online status in correct interval when online', () => {
		jest.useFakeTimers();

		fetchMock.get( healthCheckEndpoint, {
			body: healthCheckResponse,
		} );

		renderHook( () => useMonitorInternetConnection(), { registry } );

		expect( store.getState().isOnline ).toBe( true );

		// The initial fetch will happen immediately.
		expect( fetchMock ).toHaveFetchedTimes( 1 );

		jest.advanceTimersByTime( 119999 );

		expect( fetchMock ).toHaveFetchedTimes( 1 );

		jest.advanceTimersByTime( 1 );

		// The second fetch will happen after 120000ms.
		expect( fetchMock ).toHaveFetchedTimes( 2 );
	} );
```

Please can you update this test and the one below accordingly?

zutigrm replied:
Thanks, got it.

But the second test is testing the interval when use is offline, when `isOnline` is `false` it only updates the state and bails early, there is no side effect in that case, only when `isOnline` goes back online

zutigrm replied:
I was thinking, I guess absence of request is valid to test also :D

techanvil reviewed (changes requested):
Thanks @zutigrm, just one more update and this should be good to go.

zutigrm commented:
@techanvil Thanks for pointers and valuable feedback.

I updated the tests, for the last one, as mentioned [here](https://github.com/google/site-kit-wp/pull/8033#discussion_r1444971404), since there is no side effect when `isOnline` is `false`, I updated the test to confirm that fact that no request should have been made. Let mw know what you think about that

techanvil commented on assets/js/hooks/useMonitorInternetConnection.test.js:
```diff
@@ -0,0 +1,179 @@
+
+		expect( store.getState().isOnline ).toBe( false );
+
+		jest.advanceTimersByTime( 15000 );
```
I'd suggest a slightly different tactic here - in order to positively confirm the interval, we could toggle the online mode just before the interval is up. For example:

```js
	it( 'should check online status in correct interval when offline', async () => {
		jest.useFakeTimers();

		fetchMock.get( healthCheckEndpoint, {
			body: healthCheckResponse,
		} );

		mockOnlineStatus( false );

		renderHook( () => useMonitorInternetConnection(), { registry } );

		expect( store.getState().isOnline ).toBe( false );

		// The interval should be 15000ms when offline.
		jest.advanceTimersByTime( 14999 );

		// Enable online status so we can verify the interval via the fetch.
		mockOnlineStatus( true );

		// No fetch should happen when offline until the interval is reached.
		expect( fetchMock ).toHaveBeenCalledTimes( 0 );

		await act( async () => {
			jest.advanceTimersByTime( 1 );
			await new Promise( ( resolve ) => resolve() );
		} );

		expect( fetchMock ).toHaveFetchedTimes( 1 );
	} );
```

What do you think?

zutigrm replied:
That's great idea! Thanks, didn't think of that

techanvil commented on assets/js/hooks/useMonitorInternetConnection.js:
```diff
@@ -0,0 +1,94 @@
+		}
+	);
+
+	useInterval(
+		() => {
+			checkInternetConnection();
+		},
+		isOnline ? 120000 : 15000
+	);
```
We should be able to pass `checkInternetConnection` directly in here, rather than an anonymous function:

```suggestion
	useInterval( checkInternetConnection, isOnline ? 120000 : 15000 );
```

techanvil reviewed (changes requested):
> @techanvil Thanks for pointers and valuable feedback.
> 
> I updated the tests, for the last one, as mentioned [here](https://github.com/google/site-kit-wp/pull/8033#discussion_r1444971404), since there is no side effect when `isOnline` is `false`, I updated the test to confirm that fact that no request should have been made. Let me know what you think about that

Thanks @zutigrm. The only issue I see with that approach is that it's not a clear positive - I've suggested a slight variation. I've also left another minor comment elsewhere. Please see what you think.

techanvil reviewed (approved):
LGTM, nice one @zutigrm!

aaemnnosttv commented on assets/js/hooks/useMonitorInternetConnection.js:
```diff
@@ -0,0 +1,89 @@
+			// if request fails to reach the server.
+			const onlineResponse = await apiFetch( {
+				method: 'GET',
+				path: '/google-site-kit/v1/core/site/data/health-checks',
```
🚨 While this makes sense as an endpoint that we'd always have access to, it will make a request to the Search Console / Webmasters Google API every time. Together with the fact that this will be called on every page load and every few minutes or less would make this a very big increase in the number of calls we make to this API which also aren't specifically needed for this feature so it would be a substantial unnecessary increase of our quota.

If we're only concerned about the request reaching the server (as the comment above implies) we can request other endpoints or even static assets instead.

aaemnnosttv commented on assets/js/hooks/useMonitorInternetConnection.js:
```diff
@@ -0,0 +1,89 @@
+
+			// We are only interested if the request was successful, to
+			// confirm online status.
+			const canReachInternetURL = !! onlineResponse.checks;
```
`checks` is an object of the checks that were run and their results so it will always exist and be truthy if the request succeeded. If it didn't succeed, then `apiFetch` would throw instead so this line would not be reached.

aaemnnosttv commented on assets/js/hooks/useMonitorInternetConnection.js:
```diff
@@ -0,0 +1,89 @@
+		}
+
+		try {
+			// Use apiFetch directly, instead of our wrapper API, to have
+			// control over the error handling/catching, as status is switched to offline
+			// if request fails to reach the server.
+			const onlineResponse = await apiFetch( {
```
I don't quite follow the reason why we need to use `apiFetch` directly here, please explain :)

aaemnnosttv reviewed (commented):
This one needs some revision before it's ready for launch. See #8080
